first lets import the library we are going to use in this project 


In [6]:
import pandas as pd

In [9]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

i have imported pandas so I can read and change the file that contains the data in to DataFrame


In [10]:
df = pd.read_csv(r"C:\Users\rebor\Downloads\GloBOx project.csv")
df

uid   dt device  spent group     join_dt country gender
0      1000000  NaN      I    NaN     B  2023-01-28     CAN      M
1      1000001  NaN      A    NaN     A  2023-01-27     BRA      M
2      1000002  NaN      A    NaN     A  2023-02-01     FRA      M
3      1000003  NaN      I    NaN     B  2023-01-25     BRA      M
4      1000004  NaN      A    NaN     A  2023-02-04     DEU      F
...        ...  ...    ...    ...   ...         ...     ...    ...
49077  1049995  NaN      A    NaN     B  2023-02-03     BRA      F
49078  1049996  NaN      A    NaN     A  2023-01-29     USA      F
49079  1049997  NaN      A    NaN     B  2023-02-03     BRA      M
49080  1049998  NaN      I    NaN     B  2023-02-03     CAN      M
49081  1049999  NaN      I    NaN     B  2023-01-29     GBR      M

[49082 rows x 8 columns]

lets fill the null values in spent with 0 

In [ ]:
df["spent"] = df["spent"].fillna(0)

df

uid   dt device  spent group     join_dt country gender
0      1000000  NaN      I    0.0     B  2023-01-28     CAN      M
1      1000001  NaN      A    0.0     A  2023-01-27     BRA      M
2      1000002  NaN      A    0.0     A  2023-02-01     FRA      M
3      1000003  NaN      I    0.0     B  2023-01-25     BRA      M
4      1000004  NaN      A    0.0     A  2023-02-04     DEU      F
...        ...  ...    ...    ...   ...         ...     ...    ...
49077  1049995  NaN      A    0.0     B  2023-02-03     BRA      F
49078  1049996  NaN      A    0.0     A  2023-01-29     USA      F
49079  1049997  NaN      A    0.0     B  2023-02-03     BRA      M
49080  1049998  NaN      I    0.0     B  2023-02-03     CAN      M
49081  1049999  NaN      I    0.0     B  2023-01-29     GBR      M

[49082 rows x 8 columns]

now that i have the data and library lets include new dummie codes that is gone be usefull for this analysis

In [ ]:
df["conversion"] =(df["spent"] != 0 ).astype(int)
df

uid   dt device  spent group     join_dt country gender  conversion
0      1000000  NaN      I    0.0     B  2023-01-28     CAN      M           0
1      1000001  NaN      A    0.0     A  2023-01-27     BRA      M           0
2      1000002  NaN      A    0.0     A  2023-02-01     FRA      M           0
3      1000003  NaN      I    0.0     B  2023-01-25     BRA      M           0
4      1000004  NaN      A    0.0     A  2023-02-04     DEU      F           0
...        ...  ...    ...    ...   ...         ...     ...    ...         ...
49077  1049995  NaN      A    0.0     B  2023-02-03     BRA      F           0
49078  1049996  NaN      A    0.0     A  2023-01-29     USA      F           0
49079  1049997  NaN      A    0.0     B  2023-02-03     BRA      M           0
49080  1049998  NaN      I    0.0     B  2023-02-03     CAN      M           0
49081  1049999  NaN      I    0.0     B  2023-01-29     GBR      M           0

[49082 rows x 9 columns]

In [ ]:
# creating a dataset the contains value we are going to use do analysis on conversion rate
data = {
    "group" : ["A", "B"] ,
    "number_users" : [df[df["group"] == "A" ]["uid"].nunique() , df[df["group"] == "B"]["uid"].nunique()] ,
    "conversion"  : [df[df["group"] == "A"]["conversion"].sum() , df[df["group"] == "B"]["conversion"].sum()]
} 

In [ ]:
df_data = pd.DataFrame(data)
df_data

group  number_users  conversion
0     A         24402        1014
1     B         24680        1219

In [ ]:
conversion_rate_A = success_A/trial_A
conversion_rate_B = success_B/trial_B
print(conversion_rate_A,conversion_rate_B)

0.04165468512508729 0.049552845528455285


In [ ]:
# determinig the number in to field that we can use to do analysis

trial_A = df_data.loc[df_data["group"] == "A","number_users"].values[0]
trial_B = df_data.loc[df_data["group"] == "B","number_users"].values[0]
success_A = df_data.loc[df_data["group"] == "A","conversion"].values[0]
success_B = df_data.loc[df_data["group"] == "B","conversion"].values[0]
print(trial_A, trial_B, success_A, success_B)

24402 24680 1014 1219


In [ ]:
from statsmodels.stats.proportion import proportions_ztest

In [ ]:
# doing  analysis to determine if there is a significant difference between the 2 groups
success = [success_A, success_B]
trial = [trial_A, trial_B]
z_stat, p_value = proportions_ztest(success, trial)

In [ ]:
print(z_stat, p_value)

-4.1664960105779585 3.093173032978271e-05


after we have done the test we need to determine our conculusion 
lets make clear null_hypothesis 

null hypothesis : there is no significant difference in conversion rate between group A and B 
alternative hypothesis : there is significant difference in conversion rate between group A and B 

lets set α  = 0.05

after we have done  proportions z test we got a p value of 0.0000309 which is significantly smaller than the α  value of 0.05 
that indicate there is a significant difference in coversion rate , so we can reject the null hypothesis .

and 

the z test value -4.1664960105779585 indicate that there is a significantly large difference in conversion rate and the minus(-) sign represent that group B is more effective in conversion rate than group A .

so based on the numbers provided in the test and by rejecting the null value we can conclude that there a significant difference in conversion rate between group A and group B. And based on the result we got in the ztest we can conclude the difference in conversion rate is significantly large and the minus(-) sigh helps us to   clearly say group B is much better in conversion rate than group A. 





And now lets do another analysis to determine which group generates the most revenue.

In [ ]:
df

uid   dt device  spent group     join_dt country gender  conversion
0      1000000  NaN      I    0.0     B  2023-01-28     CAN      M           0
1      1000001  NaN      A    0.0     A  2023-01-27     BRA      M           0
2      1000002  NaN      A    0.0     A  2023-02-01     FRA      M           0
3      1000003  NaN      I    0.0     B  2023-01-25     BRA      M           0
4      1000004  NaN      A    0.0     A  2023-02-04     DEU      F           0
...        ...  ...    ...    ...   ...         ...     ...    ...         ...
49077  1049995  NaN      A    0.0     B  2023-02-03     BRA      F           0
49078  1049996  NaN      A    0.0     A  2023-01-29     USA      F           0
49079  1049997  NaN      A    0.0     B  2023-02-03     BRA      M           0
49080  1049998  NaN      I    0.0     B  2023-02-03     CAN      M           0
49081  1049999  NaN      I    0.0     B  2023-01-29     GBR      M           0

[49082 rows x 9 columns]

to see which group generate the most revenue we only need to consider the users that have converted. so we have to create a new dataset that only contains users that converted.



In [ ]:
df_r = df[df["spent"] > 0]
df_r

uid          dt device       spent group     join_dt country  \
38     1000039  2023-02-04      A   36.650000     B  2023-02-04     GBR   
44     1000045  2023-01-30      I   51.580000     B  2023-01-26     USA   
69     1000071  2023-02-02      I    6.710000     B  2023-01-31     USA   
98     1000101  2023-02-05      A    5.674497     B  2023-02-02     MEX   
99     1000101  2023-02-02      A   18.130000     B  2023-02-02     MEX   
...        ...         ...    ...         ...   ...         ...     ...   
48964  1049880  2023-01-26      I  186.690000     A  2023-01-26     TUR   
48975  1049891  2023-01-26      A   47.140000     B  2023-01-26     BRA   
48978  1049894  2023-01-25      I   34.050000     A  2023-01-25     BRA   
49010  1049927  2023-01-25      A   52.030000     A  2023-01-25     AUS   
49055  1049973  2023-01-31      A   85.930000     B  2023-01-27     MEX   

      gender  conversion  
38         F           1  
44         F           1  
69         F           1  
98         F           1  
99         F           1  
...      ...         ...  
48964      M           1  
48975      F           1  
48978      F           1  
49010      F           1  
49055      F           1  

[2233 rows x 9 columns]

In [ ]:
# lets create a new data set that only the values we new lets use pivot table
data_r = df_r.pivot_table( values = ["uid" , "spent"],
                          index = ["group"],
                          aggfunc = {"uid" : "count", "spent" : "sum"}
                          )

In [ ]:
data_r

spent   uid
group                    
A      82145.903065  1014
B      83415.326869  1219

In [ ]:
# adding average revenue generated by user in into the data set
data_r['avg_revenue_prr_user'] = data_r["spent"] / data_r["uid"]
data_r

spent   uid  avg_revenue  avg_revenue_prr_user
group                                                       
A      82145.903065  1014    81.011739             81.011739
B      83415.326869  1219    68.429308             68.429308

In [ ]:
# making visualizion of the normal distrubion 
from scipy.stats import norm


In [ ]:
# looking more description  in groups and spent 
more_info = df_r.groupby("group")["spent"].describe()
more_info

count       mean        std       min        25%    50%      75%  \
group                                                                      
A      1014.0  81.011739  95.935897  0.815058  38.685000  62.62  98.8175   
B      1219.0  68.429308  89.440967  0.340320  30.632592  51.16  81.8000   

          max  
group          
A      1659.4  
B      1546.3

at first glance we can see that from the people who have converted, the people in group A have generated more per user. But it could be becouse of luck(at random) we need to do statstical test to determine if there is a difference in revenue generated by each group.

In [ ]:
# making the necceseray list to make our t test
spent_A = df_r[df_r["group"] == "A"]["spent"]
spent_B = df_r[df_r["group"] == "B"]["spent"]

# importing the test we are going to use
from scipy import stats
t_stat, p_value = stats.ttest_ind(spent_A, spent_B, equal_var=False)  
print(t_stat, p_value)

3.1816934886425643 0.0014855935907930756


Now we have done our analysis lets make our null hypothesis 
null hypothesis : there is no difference in revenue generated by each group 
alternate hypothesis : there is a difference in revenue generated by each group 
 
lets α  = 0.05

using the p value we got in the test we did that is 0.00145, and when we compare it with α  value 0f 0.05 it is significantly smaller. which help me to conclude to reject the null hypothesis. so we can conclude there is a difference in revenue generated by each group

and the t_test will help us to conclude to determine which group is better at generating revenue and if the difference is big enough

the t_test value of 3.1817 tell us that the difference in revenue generated is large enough to conclude it is not becouse of random.
and (+) positive tells us which group is better in this case. Group A is better at generating revenue.

To conclude there is a differnce in revenue generated by group A & B . The difference in the revenue generation is significantly large and Group A is much better than group B. 

So our data tells the story of 2 different ways of advertising  that have their own advantages. The first one is not that good in conversion rate compared to the second one but has much more revenue generation rate. The second one has a much better rate in conversion rate but lucks the revenue generation rate of first one. 